src/site/pages/digital/tests/containers/ExprToCircuitPopup.test.tsx

```diff
@@ -0,0 +1,143 @@
+    const user = userEvent.setup();
+
+    beforeEach(() => {
+        render(<Provider store={store}><ExprToCircuitPopup mainInfo={info} /></Provider>);
+        act(() => {store.dispatch(OpenHeaderPopup("expr_to_circuit"))});
+    });
+
+    afterEach(() => {
+        info.designer.reset();
+    });
+
+    test("Popup Created with default states", () => {
+        // Check header and button states
+        expect(screen.getByText("Digital Expression To Circuit Generator")).toBeVisible();
+        expect(screen.getByText("Cancel")).toBeVisible();
+        expect(screen.getByText("Generate")).toBeVisible();
+        expect(screen.getByText("Generate")).toBeDisabled();
+
+        // Check format options
+        const [onButton1, offButton1] = GetToggles(/Programming 1/);
+        expect(onButton1).toBeVisible();
+        expect(offButton1).not.toBeVisible();
+        const [onButton2, offButton2] = GetToggles(/Custom/);
+        expect(onButton2).not.toBeVisible();
+        expect(offButton2).toBeVisible();
+        expect(screen.queryByText(/Custom AND/)).toBeNull();
+
+        // Check toggle switches
+        const [onButton3, offButton3] = GetToggles(/Place labels for inputs/);
+        expect(onButton3).not.toBeVisible();
+        expect(offButton3).toBeVisible();
+        const [onButton4, offButton4] = GetToggles(/Generate into IC/);
+        expect(onButton4).not.toBeVisible();
+        expect(offButton4).toBeVisible();
+        expect(screen.queryByText(/Connect Clocks/)).toBeNull();
+
+        // Check dropdowns
+        const inputOptions = screen.getByLabelText(/Input Component/).querySelectorAll("option");
+        const switchInputOption = [...inputOptions].find((input) => input.text === "Switch");
+        expect(switchInputOption?.selected).toBeTruthy();
+        const outputOptions = screen.getByLabelText(/Output Component/).querySelectorAll("option");
+        const ledOutputOption = [...outputOptions].find((input) => input.text === "LED");
+        expect(ledOutputOption?.selected).toBeTruthy();
+
+        // Text input is empty
+        const input = screen.getByPlaceholderText("!a | (B^third)") as HTMLInputElement;
+        expect(input.value).toBe("");
+    });
+
+    test("Cancel Button Cancels", async () => {
+        await user.type(screen.getByPlaceholderText("!a | (B^third)"), "a | b");
+
+        await user.click(screen.getByText("Cancel"));
+        expect(screen.getByText("Cancel")).not.toBeVisible();
+        expect(screen.getByText("Digital Expression To Circuit Generator")).not.toBeVisible();
+
+        // Reopen and requery in case reference changed
+        act(() => {store.dispatch(OpenHeaderPopup("expr_to_circuit"))});
+        expect((screen.getByPlaceholderText("!a | (B^third)") as HTMLInputElement).value).toBe("");
+    });
+
+    test("Generate Button", async () => {
+        // Enter the expression and generate
+        await user.type(screen.getByPlaceholderText("!a | (B^third)"), "a | b");
```

Comment:
that's a gross way to `get` it by? any better way?

Reply:
Turns out there was a better way, I didn't think getByRole would work since it isn't a div but it did.